src/libostree/ostree-mutable-tree.c

```diff
@@ -47,10 +47,24 @@ struct OstreeMutableTree
 {
   GObject parent_instance;
 
+  /* The parent directory to this one.  We don't hold a ref because this mtree
+   * is owned by the parent.  We can be certain that any mtree only has one
+   * parent because external users can't set this, it's only set when we create
+   * a child from within this file (see insert_child_mtree). We ensure that the
+   * parent pointer is either valid or NULL because when the parent is destroyed
+   * it sets parent = NULL on all its children (see remove_child_mtree) */
+  OstreeMutableTree *parent;
```

Comment:
This sounds like it might be a good fit for weak refs: https://developer.gnome.org/gobject/stable/gobject-The-Base-Object-Type.html#g-object-add-weak-pointer. So then your hash table value free function can just be `g_object_unref`.

Reply:
I did consider this.  I figured it's not worth the overhead both in the source and at runtime.  The difference would be:

* Atomic instructions would be used for the weak refcnt.  It isn't threadsafe to use different `OstreeMutableTree` instances that are part of the same hierarchy from different threads anyway as we have no synchronisation around `contents_checksum`.  We could add more synchronisation here but it hardly seems worth it.
* The (immediate) parents would not be freed until the children had been destroyed: the weak pointer still needs to point to somewhere to check the weak refcnt (assuming it's similar to C++'s `shared_ptr` with `make_shared`)

They're minor points but I think the current implementation is nicer and we avoid a bunch of error handling for codepaths that will never be taken.  I'm not wed to the current implementation so I can change it to use weak_refs if you'd like.

Reply:
This is just a quick comment, I haven't fully reviewed this patch yet.

Yeah, GLib's weak refs use a global mutex.  I don't know if that really matters here (probably not) - bigger perf wins would probably be having a more efficient string-tree structure.

Having non-owning pointers for trees is pretty standard I believe.

Reply:
I just looked at Rust's BTree for example, and there the [parent pointer](https://github.com/rust-lang/rust/blob/9f79d2f86af73da483f4fe8e2422c15c79221d32/src/liballoc/btree/node.rs#L71) is a raw pointer.  It kind of has to be in Rust of course, otherwise it'd be cyclic.